### PR TITLE
Feat: basic ui tests

### LINKED ISSUE #41: Write the basic front-end UI tests
At the moment there is only 2 UI tests using spectron and mocha. They act as a simple smoke test.

Need a lot more tests covering the basic usage scenarios of the application via the UI. Write tests for:

- [ ] Going to the user settings, adding an ethereum account and looking at its balance
- [ ] Adding a token and seeing that tokens get updated
- [ ] Adding a bitcoin account and seeing that the balance is there
- [ ] Adding a fiat balance and see that it's properly updated
- [ ] Add an exchange and see that the API key is succesfully accepted
- [ ] Change the currency to EUR and that the currency changes everywhere in the UI

### PR DESCRIPTION
**Pull request by mykeels**
fixes #41 

### 6 Tests Written

- [x] Going to the user settings, adding an ethereum account and looking at its balance
- [x] Adding a token and seeing that tokens get updated
- [x] Adding a bitcoin account and seeing that the balance is there
- [x] Adding a fiat balance and see that it's properly updated
- [x] Add an exchange and see that the API key is succesfully accepted
- [x] Change the currency to EUR and that the currency changes everywhere in the UI

### Pending

- [ ] Change each one of the settings and see that the change has been applied

The pending test case is unclear ... Could you give examples of each setting you expect to the tested, and what effect the change is to make?

### CODE REVIEW

**Comment by mykeels** [comment]
@LefterisJP what version of node is used in travis-ci?

**Comment by LefterisJP** [comment]
@mykeels I supposed v8.9.1.

You can see available versions at the start: https://travis-ci.org/rotkehlchenio/rotkehlchen/jobs/421756184#L402

**Comment by LefterisJP** [comment]
@mykeels Thank you for your work so far. But please don't mess with the python dependencies and please don't add other travis jobs. The issue is only on the front-end. Also by the end of this issue I expect a maximum of 6 commits, so all the commits you are making should be squashed.

**Comment by mykeels** [comment]
@LefterisJP I'm having trouble getting travis to build. First, it was because a Node version less than 8.x was being used on travis. I've made sure to install 8.9 as given in the source_install.md file.

Now, I expect it's an issue with the python version, which should be 3.6, but was 2.7 in travis. Should I ignore these?

**Comment by LefterisJP** [comment]
@mykeels Hmmm I see. No don't ignore it. Python should be 3.6. Why was it 2.7?

Perhaps because the mocha tests are not ran properly inside the python venv? I will also have a look.

File: .travis.yml [hunk 1]
```diff
@@ -54,12 +61,12 @@ script:
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then source terryfy/test_tools.sh; fi
   - if [[ "$TRAVIS_OS_NAME" == "osx" ]]; then python -m pytest rotkehlchen; fi
   - if [[ "$TRAVIS_OS_NAME" != "osx" ]]; then coverage run --source rotkehlchen/ -m pytest --travis-fold=always $TEST_TYPE; fi
-  - npm config set python python2.7
+  - npm config set python python3.6
```
Comment (LefterisJP) [inline .travis.yml]: There is no need to change this. The reason I set npm's python here is only for electron rebuild. This has nothing to do with the python used in the tests.

File: .travis.yml [hunk 2]
```diff
@@ -43,6 +45,11 @@ before_install:
     fi
 
 install:
+  - pip install virtualenv
```
Comment (LefterisJP) [inline .travis.yml]: All of these should not be here. Travis is already creating a venv for us. I think that you adding `language: 'node_js'` above may have changed this. I would strongly prefer it you remove the `language: 'node_js' from above and find another way to specify a node_js version in Travis that does not force you to add all the stuff about python you added.

File: main.js [hunk 3]
```diff
@@ -94,8 +104,11 @@ const createPyProc = () => {
         pyProc = require('child_process').spawn('python', ["-m", "rotkehlchen", "--zerorpc-port", port]);
     }
 
+    pyProc.on('data', (data) => {
```
Comment (LefterisJP) [inline main.js]: Why did you add this callback? What does it do?

File: main.js [hunk 4]
```diff
@@ -5,6 +5,16 @@ const BrowserWindow = electron.BrowserWindow;
 const path = require('path');
 const fs = require('fs');
 
+if (process.env.NODE_ENV != 'production') {
```
Comment (LefterisJP) [inline main.js]: What is this?

File: package.json [hunk 5]
```diff
@@ -12,7 +12,9 @@
     "devDependencies": {
         "chai": "^4.1.2",
         "chai-as-promised": "^7.1.1",
+        "dotenv": "^6.0.0",
         "electron": "1.8.4",
+        "electron-context-menu": "^0.10.0",
```
Comment (LefterisJP) [inline package.json]: What are these two new dependencies for?

File: ui/user_settings.js [hunk 6]
```diff
@@ -362,6 +362,7 @@ function add_blockchain_account(event) {
             if (blockchain == 'ETH') {
                 recreate_ethchain_per_account_table(result['per_account']['ETH']);
             } else if (blockchain == 'BTC') {
+                create_blockchain_balances_tables(result)
```
Comment (LefterisJP) [inline ui/user_settings.js]: I don't think this is correct here. Why did you add it?

File: ui/userunlock.js [hunk 7]
```diff
@@ -97,7 +97,7 @@ function prompt_sign_in() {
         buttons: {
             formSubmit: {
                 text: 'Sign In',
-                btnClass: 'btn-blue',
+                btnClass: 'btn-blue sign-in',
```
Comment (LefterisJP) [inline ui/userunlock.js]: Why add `sign-in` in the class here? What does it achieve?

File: ui/userunlock.js [hunk 8]
```diff
@@ -109,7 +109,7 @@ function prompt_sign_in() {
             },
             newAccount: {
                 text: 'Create New Account',
-                btnClass: 'btn-blue',
+                btnClass: 'btn-blue create-new-account',
```
Comment (LefterisJP) [inline ui/userunlock.js]: Why add `create-new-account` in the class here? What does it achieve?

**Comment by LefterisJP** [comment]
Also @mykeels 

> Change each one of the settings and see that the change has been applied

I can help with that part of testing. If you can't finish this particular part of the task alone it's fine.

File: ui/userunlock.js (hunk 8 above)
Comment (mykeels) [reply ui/userunlock.js]: The `create-new-account` and `sign-in` classes are used to identify the buttons, so they can be referenced and clicked during tests. Do you think an `id` attribute would be more appropriate?

File: ui/userunlock.js (hunk 7 above)
Comment (mykeels) [reply ui/userunlock.js]: https://github.com/rotkehlchenio/rotkehlchen/pull/136#discussion_r213509599

File: ui/user_settings.js (hunk 6 above)
Comment (mykeels) [reply ui/user_settings.js]: The table for displaying the BTC balances was not showing up, and a null reference error was thrown in console when the BTC address was added.

Looking through the code for the `create_blockchain_balances_tables`, I saw that it created the element, and handler for the BTC balances table.

Placing it here fixed the issue. Do you have a better suggestion?

File: package.json (hunk 5 above)
Comment (mykeels) [reply package.json]: The `electron-context-menu` module gives the `inspect element` right-click functionality, in the electron app, only in `development` mode. 

It helped me with debugging. I could take it out if you're uncomfortable with it.

File: main.js (hunk 4 above)
Comment (mykeels) [reply main.js]: https://github.com/rotkehlchenio/rotkehlchen/pull/136#discussion_r213510227

File: main.js (hunk 3 above)
Comment (mykeels) [reply main.js]: I'm taking this out

File: .travis.yml (hunk 2 above)
Comment (mykeels) [reply .travis.yml]: Okay, doing that now ...

**Comment by mykeels** [comment]
About https://github.com/rotkehlchenio/rotkehlchen/pull/136#issuecomment-416767013 I can finish the task, I just need it to be clear.

**Comment by mykeels** [comment]
Looks like [travis is building now](https://travis-ci.org/rotkehlchenio/rotkehlchen/jobs/421833959), but the tests are failing. You'd need to add environment variables to the travis config. See the environment variables to be added below:

| Name | Description |
| ------ |:---:|
|  PASSWORD  | A password that user accounts created for test purposes would have |
|  ETH_ADDRESS  | An actual ETH Address for testing for ETH balances |
|  BTC_ADDRESS  | An actual BTC Address for testing for BTC balances |
|  BITTREX_API_KEY  | An actual Bittrex API Key |
|  BITTREX_API_SECRET  | An actual Bittrex API Secret |

If testing locally, you'd want to create an `.env` file in the application's root directory and provide these details in the following format:

```
PASSWORD=
ETH_ADDRESS=
BTC_ADDRESS=
BITTREX_API_KEY=
BITTREX_API_SECRET=
```

**Comment by LefterisJP** [comment]
@mykeels Thank you very much for your work. So when I find some time I will review your PR and try to test it.

> https://github.com/rotkehlchenio/rotkehlchen/pull/136#issuecomment-416781085

Great. Let me read it and see what's unclear. More settings have been added since I made the issue and some of them may be a bit hard to test for. Let me check and see what makes sense to test and what does not.

File: ui/userunlock.js (hunk 8 above)
Comment (LefterisJP) [reply ui/userunlock.js]: yeah I guess id would be better, since the `btnClass` is used by the pop up module I am using just to determine how the button looks.

File: package.json (hunk 5 above)
Comment (LefterisJP) [reply package.json]: I see. I will test it when I test your branch, but if it's just a development dependency it should be fine.

File: ui/userunlock.js (hunk 8 above)
Comment (mykeels) [reply ui/userunlock.js]: @LefterisJP what property here would translate to an `id` attribute?

**Comment by LefterisJP** [comment]
Hello @mykeels 

I am testing locally now and I see some flakiness. I get a lot of times the following:

```
Error: unknown error: Element <button id="add_account_button" type="submit" class="btn btn-default">...</button> is not clickable at point (323, 439). Other element would receive the click: <div class="jconfirm-box jconfirm-hilight-shake jconfirm-type-green jconfirm-type-animated" role="dialog" aria-labelledby="jconfirm-box24877" tabindex="-1" style="transition-duration: 0.4s; transition-timing-function: cubic-bezier(0.36, 0.55, 0.19, 1);">...</div>
```

Seems that sometimes the jconfirm modal box does not go away fast enough. Any idea of the best way to get rid of it?


Also please note that I am going to push some fixes and rebase your branch on master.

Finally in order for me to accept your commits I will need you to sign the CLA that has popped up as a comment in this repo.

**Comment by LefterisJP** [comment]
Locally they all pass, apart from the flakiness I spoke about above.

But as you can see from [here](https://travis-ci.org/rotkehlchenio/rotkehlchen/jobs/423371609#L837) in Travis `npm install` fails with 

```
npm ERR! request to https://registry.npmjs.org/datatables.net failed, reason: Hostname/IP doesn't match certificate's altnames: "Host: registry.npmjs.org. is not in the cert's altnames: DNS:a.sni.fastly.net, DNS:a.sni.global-ssl.fastly.net"
```

which is a rather strange errror

**Comment by LefterisJP** [comment]
Hey @mykeels I have pushed some commits to adjust travis and improve the tests. I tried to address the flakiness and also combined the eth test with the token test, because the token needs to be tested after a user enters an ETH account.

You still need to:
1. sign the CLA.
2. In OSX tests we get a syntax error for `npm test`
3. In Linux in Travis we got lots of flakiness. Do you know  of any way to reduce it? Locally some tests are also flaky.
4. Regarding the settings test I still need to think about it a bit, but 1-3 are a much higher priority.

**Comment by mykeels** [comment]
@LefterisJP perhaps adding pauses between the statements would help make sure the `.jconfirm` box goes away before the next action is performed.

Also, I see you have added the `.env` file to source control. Is there a reason you don't wish to use Travis's and AppVeyor's Environment Variable setting functionality?

**Comment by LefterisJP** [comment]
> Also, I see you have added the .env file to source control. Is there a reason you don't wish to use Travis's and AppVeyor's Environment Variable setting functionality?

Which functionality do you mean? Is it this one: https://docs.travis-ci.com/user/environment-variables/

If yes I have also added them as environment variables there but I realized now that it's superfluous. What are the pros/cons? All these are test variables so nothing sensitive.

File: ui/tests/add-btc.spec.js [hunk 9]
```diff
@@ -48,12 +48,12 @@ describe('User Settings', function () {
     await this.app.client.click('.jconfirm-buttons>button')
 
     // wait for popup modal, then close it
-    await this.app.client.waitForExist('.jconfirm-box.jconfirm-type-green.jconfirm-type-animated', 5000)
+    await this.app.client.waitForExist('.jconfirm-box', 5000)
     await this.app.client.execute(function () {
         $('.jconfirm').remove()
     })
     // wait for the other modal popup, then close it
-    await this.app.client.waitForExist('.jconfirm-box.jconfirm-hilight-shake.jconfirm-type-animated.jconfirm-type-green', 5000)
```
Comment (LefterisJP) [inline ui/tests/add-btc.spec.js]: I may be wrong, but I put the full class path for a reason. The order with which these popups appear is not guaranteed and I needed a unique way to address each one. So I think the full class path should be there.

Also @mykeels do you know of any way to abstract this functionality which we use in every tests somehow? Would normal functions work here?

**Comment by mykeels** [comment]
I was referring to [this](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings) @LefterisJP because it does contain Bittrex API Key and Secret

**Comment by LefterisJP** [comment]
@mykeels I know it's an empty account I created just for testing. Has nothing in there.

**Comment by mykeels** [comment]
@LefterisJP what could make the backend app crash immediately it is started? I keep getting 

```
CREATED PYPROCESS
python subprocess killed with signal null and code 1
```

**Comment by LefterisJP** [comment]
Where do you get it? Running it locally? Have you properly made the python environment as described in the source installation [docs](https://github.com/rotkehlchenio/rotkehlchen/blob/master/docs/source_install.md)?

A way to test this is to run the backend alone. Assuming you have everything setup properly run `python -m rotkehlchen` and see what happens.

**Comment by mykeels** [comment]
You're right, I'd forgotten to run `source ./bin/activate` to make sure python3 was the default python version. Is there some similar setting we have to do on travis to ensure python3 is used as the default? @LefterisJP

**Comment by LefterisJP** [comment]
> Is there some similar setting we have to do on travis to ensure python3 is used as the default? @LefterisJP

Travis uses python3 for rotkehlchen. It's in [travis.yml](https://github.com/rotkehlchenio/rotkehlchen/blob/bbbd349b8850e03c38b472e93f2810e3202851c3/.travis.yml#L5)

**Comment by mykeels** [comment]
@LefterisJP travis is building successfully. Is it possible to re-run the AppVeyor build without pushing changes?

**Comment by LefterisJP** [comment]
@mykeels I can do it as many times as you need, yes.

But it still looks like a flaky test:

```
 Adding a token and seeing that tokens get updated:
     Error: stale element reference: element is not attached to the page document
      at new RuntimeError (node_modules\webdriverio\build\lib\utils\ErrorHandler.js:143:12)
      at Request._callback (node_modules\webdriverio\build\lib\utils\RequestHandler.js:316:39)
      at Request.self.callback (node_modules\request\request.js:185:22)
      at Request.<anonymous> (node_modules\request\request.js:1161:10)
      at IncomingMessage.<anonymous> (node_modules\request\request.js:1083:12)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

**Comment by LefterisJP** [comment]
Closing since this was already implemented by: https://github.com/rotkehlchenio/rotkehlchen/pull/207
